### PR TITLE
cache: clear snapshot files from cache during load index

### LINKED ISSUE #4934: Cleanup removed snapshots from cache
Output of `restic version`
--------------------------
0.16.5

What should restic do differently? Which functionality do you think we should add?
----------------------------------------------------------------------------------
When using a restic repository from multiple host, then after running `forget` on one of the hosts, the removed snapshots still exist in the cache of the other hosts. Restic should remove those snapshots from the cache of all hosts.

It would be sufficient, if for example the commands that list all snapshots, also automatically clean up the cache.
For that it might be possible to hook into the `List` function of a `Repository` to remove stale snapshots from the cache.

What are you trying to do? What problem would this solve?
---------------------------------------------------------
Keep the cache size small.

### PR DESCRIPTION
**Pull request by konidev20**
What does this PR change? What problem does it solve?
-----------------------------------------------------
clear snapshot files from cache during repository preparation. 

Was the change previously discussed in an issue or on the forum?
----------------------------------------------------------------
There is a function in `repository.go` called `prepareCache()` which is used to clean up old index and pack files from the local cache. This method is called during `LoadIndex`.

Closes #4934

Checklist
---------
- [x] I have read the [contribution guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches).
- [x] I have [enabled maintainer edits](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [x] I have added tests for all code changes.
- [ ] I have added documentation for relevant changes (in the manual).
- [x] There's a new file in `changelog/unreleased/` that describes the changes for our users (see [template](https://github.com/restic/restic/blob/master/changelog/TEMPLATE)).
- [x] I have run `gofmt` on the code in all commits.
- [x] All commit messages are formatted in the same style as [the other commits in the repo](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits).
- [x] I'm done! This pull request is ready for review.

### CODE REVIEW

File: internal/repository/repository.go [hunk 1]
```diff
@@ -723,6 +723,21 @@ func (r *Repository) prepareCache() error {
...
+		return nil
+	})
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "error listing snapshots: %v\n", err)
+	}
+
+	err = r.Cache.Clear(restic.SnapshotFile, snapshots)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "error clearing snapshot files in cache: %v\n", err)
+	}
```
Comment (konidev20) [inline internal/repository/repository.go:739]: I realised I can do the same thing, where ever we are listing the snapshots, as mentioned in the original issue.

Comment (MichaelEischer) [reply internal/repository/repository.go:739]: Explicitly listing the snapshots is somewhat cleaner, but requires an extra List call that must be avoided.

I can imagine two implementation variants: either hijack Repository.List or maybe better add a List method to the cache backend and let it take care of cleaning up indexes and snapshots itself. The latter variant has the benefit of keeping most of the cache management within the cache itself.

**Review by MichaelEischer — CHANGES_REQUESTED** [review]
I don't want to introduce an extra List call to backends just to clean up old snapshots.

File: internal/backend/cache/backend.go [hunk 2]
```diff
@@ -215,3 +218,40 @@ func (b *Backend) IsNotExist(err error) bool {
...
+		return fn(f)
+	}
+
+	err := b.Backend.List(ctx, t, wrapFn)
+	if err != nil {
+		return err
+	}
+
+	// clear the cache for files that are not in the repo anymore, ignore errors
+	err = b.Cache.Clear(t, ids)
```
Comment (MichaelEischer) [inline internal/backend/cache/backend.go:251]: Please check that the `ctx.Err() != nil` before clearing the cache.

File: internal/backend/cache/backend.go [hunk 3]
```diff
@@ -215,3 +218,40 @@ func (b *Backend) IsNotExist(err error) bool {
 func (b *Backend) Unwrap() backend.Backend {
 	return b.Backend
 }
+
+func (b *Backend) List(ctx context.Context, t backend.FileType, fn func(f backend.FileInfo) error) error {
```
Comment (MichaelEischer) [inline internal/backend/cache/backend.go:222]: This change makes the `err := r.Cache.Clear(restic.IndexFile, indexIDs)` call in `internal/repository/repository.go` redundant.

**Review by MichaelEischer — COMMENTED** [review]
The approach looks fine in general, I just have a few small remarks on it. Other than that only a test and changelog entry are missing.

File: internal/repository/repository.go [hunk 4]
```diff
@@ -699,33 +693,6 @@ func (r *Repository) createIndexFromPacks(ctx context.Context, packsize map[rest
...
-	}
-
-	indexIDs := r.idx.IDs()
-	debug.Log("prepare cache with %d index files", len(indexIDs))
-
-	// clear old index files
-	err := r.Cache.Clear(restic.IndexFile, indexIDs)
-	if err != nil {
-		fmt.Fprintf(os.Stderr, "error clearing index files in cache: %v\n", err)
-	}
```
Comment (MichaelEischer) [inline internal/repository/repository.go:716]: why did you remove the whole `prepareCache` function? All pack files are only listed by `check` or `prune`, so we can't remove the step that uses the index to determine which pack files should still be kept. Please revert those changes and _only_ remove the code between line 709 and 716.

File: changelog/unreleased/issue-4934 [hunk 5]
```diff
@@ -0,0 +1,9 @@
+Enhancement: Clear removed snapshots, index and pack files from the local cache
```
Comment (MichaelEischer) [inline changelog/unreleased/issue-4934:1]: `index` and `pack` files already get removed from the cache before this PR. Thus the should not be mentioned in the changelog entry.

File: changelog/unreleased/issue-4934 [hunk 6]
```diff
@@ -0,0 +1,9 @@
...
+
+Restic did not clear removed snapshots from the cache after the `forget`
```
Comment (MichaelEischer) [inline changelog/unreleased/issue-4934:3]: That remark is too broad. `forget` removes snapshots from the cache, but only on the current host. Please write something like the following:

```
Restic only removed snapshots from the cache on the host that runs the `forget` command.
On other hosts that use the same repository, the old snapshots remained in the cache.

Now, restic automatically clears old snapshots.
```

**Review by MichaelEischer — CHANGES_REQUESTED** [review]

File: internal/repository/repository.go (hunk 4 above)
Comment (konidev20) [reply internal/repository/repository.go:716]: I have reverted this change. I have removed *only* lines between 709 and 716.

**Review by MichaelEischer — APPROVED** [review]
LGTM. Thanks!
